runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala

```diff
@@ -0,0 +1,102 @@
+
+final case class GradingSummary(score: Int, maxScore: Int, feedback: String)
+object GradingSummary {
+  class StringSplitExtractor(splitter: String) {
+    def unapplySeq(str: String) = Array.unapplySeq(str.split(splitter))
+  }
+  object ToInt {
+    def unapply(s: String) = Try(s.toInt).toOption
+  }
+
+  /** suiteName::suiteWeight */
+  object SuiteId extends StringSplitExtractor("::")
+
+  /** suiteName::testName::testWeight */
+  object TestId extends StringSplitExtractor("::")
+
+  final case class Suite(val name: String, val weight: Int,
+    var complete: Boolean = false,
+    val tests: mutable.Map[String, Test] = mutable.Map.empty)
+
+  final case class Test(val name: String, val weight: Int,
+    var failure: Option[String])
+
+  /** Given test completion state, compute the grade and provide feedback. */
+  def apply(suites: List[Suite]): GradingSummary = {
+    val score = suites.map { _.tests.values.map { t => t.failure.fold(t.weight)(_ => 0) }.sum }.sum
+    val maxScore = suites.map { _.weight }.sum
+    val feedback = {
+      val sb = new StringBuilder
+      sb.append {
+        "Your solution achieved a testing score of %d out of %d.\n\n".format(score, maxScore)
+      }
+      if (score == maxScore)
+        sb.append("Great job!!!\n\n")
+      else {
+        sb.append("""Below you can see a short feedback for every test that failed,
+                    |indicating the reason for the test failure and how many points
+                    |you lost for each individual test.
+                    |
+                    |Tests that were aborted took too long too complete or crashed the
+                    |JVM. Such crashes can arise due to infinite non-terminitaing
+                    |loops or recursion (StackOverflowException) or excessive mamory
+                    |consumption (OutOfMemoryException).
+                    |
+                    |""".stripMargin)
+        for {
+          s <- suites
+          t <- s.tests.values
+          msg <- t.failure
+        } {
+          sb.append(s"[Test Description] ${t.name}\n")
+          sb.append(s"[Observed Error] $msg\n")
+          sb.append(s"[Lost Points] ${t.weight}\n\n")
+        }
+      }
+      sb.toString
+    }
+    GradingSummary(score, maxScore, feedback)
+  }
+
+  /** Replay the event records from the file and construct test completion state. */
+  def apply(file: File): GradingSummary = {
+    val jsons = io.Source.fromFile(file).getLines.mkString("\n").split("\n\n")
+    val entries = jsons.map(_.unpickle[Entry])
+    val suites = mutable.Map.empty[String, Suite]
+    entries.foreach {
+      case SuiteStart(SuiteId(name, ToInt(weight))) =>
+        suites += name -> Suite(name, weight)
+      case SuiteEnd(SuiteId(name, _)) =>
+        suites(name).complete = true
+      case TestStart(TestId(suiteName, name, ToInt(weight))) =>
+        suites(suiteName).tests +=
+          name -> Test(name, weight, failure = Some("test has been aborted"))
```

Comment:
@vjovanov By default tests have "test has been aborted" failure until matching TestSuccess or TestFailure happens. This means that there is no need for extra post-processing to weed out aborted tests.